app/src/main/java/org/oppia/android/app/mydownloads/downloads/DownloadsFragmentPresenter.kt

```diff
@@ -0,0 +1,253 @@
+        when (downloadsItemViewModel) {
+          is DownloadsSortByViewModel -> ViewType.VIEW_TYPE_SORT_BY
+          is DownloadsTopicViewModel -> ViewType.VIEW_TYPE_TOPIC
+          else -> throw IllegalArgumentException(
+            "Encountered unexpected view model: $downloadsItemViewModel"
+          )
+        }
+      }
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_SORT_BY,
+        inflateDataBinding = DownloadsSortbyBinding::inflate,
+        setViewModel = this::bindDownloadsSortBy,
+        transformViewModel = { it as DownloadsSortByViewModel }
+      )
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_TOPIC,
+        inflateDataBinding = DownloadsTopicCardBinding::inflate,
+        setViewModel = this::bindDownloadsTopicCard,
+        transformViewModel = { it as DownloadsTopicViewModel }
+      )
+      .build()
+  }
+
+  private enum class ViewType {
+    VIEW_TYPE_SORT_BY,
+    VIEW_TYPE_TOPIC
+  }
+
+  private fun bindDownloadsSortBy(
+    binding: DownloadsSortbyBinding,
+    viewModel: DownloadsSortByViewModel
+  ) {
+    binding.viewModel = viewModel
+    val sortByItemsList = mutableListOf<String>()
+    SortByItems.values().forEach {
+      sortByItemsList.add(fragment.getString(it.value))
+    }
+    val sortItemAdapter = ArrayAdapter(
+      fragment.requireContext(),
+      R.layout.downloads_sortby_menu,
+      sortByItemsList
+    )
+    // setting input type to zero makes AutoCompleteTextView no editable
+    binding.sortByMenu.inputType = 0
+    binding.sortByMenu.setText(
+      sortItemAdapter.getItem(previousSortTypeIndex).toString(),
+      /* filter= */ false
+    )
+    binding.sortByMenu.setAdapter(sortItemAdapter)
+
+    binding.sortByMenu.setOnItemClickListener { parent, _, position, _ ->
+      if (previousSortTypeIndex != position) {
+        lateinit var sortedDownloadsItemViewModel: MutableList<DownloadsItemViewModel>
+        when (parent.getItemAtPosition(position)) {
+          fragment.getString(SortByItems.NEWEST.value) -> {
+            // TODO(#552): update it with the time stamp value in the list
+            sortedDownloadsItemViewModel = sortTopicDownloadSize()
+          }
+          fragment.getString(SortByItems.ALPHABETICAL.value) -> {
+            sortedDownloadsItemViewModel = sortTopicAlphabetically()
+          }
+          fragment.getString(SortByItems.DOWNLOAD_SIZE.value) -> {
+            sortedDownloadsItemViewModel = sortTopicDownloadSize()
+          }
+        }
+        previousSortTypeIndex = position
+        sortByListIndexListener.onSortByItemClicked(previousSortTypeIndex)
+        binding.sortByMenu.setText(
+          sortItemAdapter.getItem(position).toString(),
+          /* filter= */ false
+        )
+        downloadsRecyclerViewAdapter.setData(sortedDownloadsItemViewModel)
+      }
+    }
+  }
+
+  private fun sortTopicAlphabetically(): MutableList<DownloadsItemViewModel> {
```

Comment:
@aggarwalpulkit596 any suggestion on how we can minimize this by using kotlin functions?

Reply:
will have to see this locally once and see what all i can minimise